File: 698.划分为k个相等的子集.cpp
/*
 * @lc app=leetcode.cn id=698 lang=cpp
 *
 * [698] 划分为k个相等的子集
 */

// @lc code=start
#include<vector>
#include<algorithm>
#include<numeric>
using namespace std;
class Solution {
public:
    vector<int>* v;
    int *arr,_k;
    bool valid(int ind, int limit){
        if(ind==v->size())return true;
        for(int i=0;i<_k;++i){
            if((*v)[ind]+arr[i]<=limit){
                arr[i]+=(*v)[ind];
                if(valid(ind+1,limit))return true;
                arr[i]-=(*v)[ind];
            }
            if(arr[i]==0||arr[i]+(*v)[ind]==limit)return false;
        }
        return false;
    }
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        sort(nums.begin(),nums.end(),greater<int>());
        v=&nums;
        _k=k;
        arr=new int[k];
        for(int i=0;i<k;++i)arr[i]=0;
        return valid(0,accumulate(nums.begin(),nums.end(),0)/k);
    }
};
// @lc code=end

File: 698.划分为k个相等的子集.py
#
# @lc app=leetcode.cn id=698 lang=python3
#
# [698] 划分为k个相等的子集
#

# @lc code=start
from typing import List, Tuple
from itertools import combinations
def mycache(dd):
    def _mycache(func):
        def ret(self,x):
            if x in dd:return dd[x]
            tmp=func(self,x)
            dd[x]=tmp
            return tmp
        return ret
    return _mycache
class Solution:
    dd={}
    def subsets(self, nums: Tuple[int]) -> List[Tuple[int]]:
        return [j for i in range(len(nums)) for j in combinations(nums,i)]
    def Tuplesub(self,a:Tuple[int],b:Tuple[int])->Tuple[int]:
        ret=list(a)
        for i in b:ret.remove(i)
        return tuple(ret)
    @mycache(dd)
    def func(self,nums: Tuple[int]):
        if not nums:return True
        if sum(nums)==self.val:return True
        for i in self.subsets(nums):
            if sum(i)==self.val:
                if self.func(self.Tuplesub(nums,i)):
                    return True
        return False
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        total=sum(nums)
        if total%k:return False
        self.val=total//k
        nums.sort(reverse=True)
        if nums[0]>self.val:return False
        self.dd.clear()
        return self.func(tuple(nums))

# @lc code=end